protocols/gossipsub/src/behaviour/tests.rs

```diff
@@ -6053,3 +6056,390 @@ fn test_priority_messages_are_always_sent() {
+    );
+    assert_eq!(
+        gs.mesh.get(&topic_hashes[1]).unwrap().len(),
+        4,
+        "topic2 should maintain 4 peers after heartbeat"
+    );
+    assert_eq!(
+        gs.mesh.get(&topic_hashes[2]).unwrap().len(),
+        6,
+        "topic3 should maintain 6 peers after heartbeat"
+    );
+
+    // Unsubscribe from topic1
+    assert!(
+        gs.unsubscribe(&Topic::new(topic_hashes[0].to_string())),
+        "Should unsubscribe successfully"
+    );
+
+    // verify it's removed from mesh
+    assert!(
+        !gs.mesh.contains_key(&topic_hashes[0]),
+        "topic1 should be removed from mesh after unsubscribe"
+    );
+
+    // re-subscribe to topic1
+    assert!(
+        gs.subscribe(&Topic::new(topic_hashes[0].to_string()))
+            .unwrap(),
+        "Should subscribe successfully"
+    );
+
+    // Verify mesh is recreated with correct size
+    assert_eq!(
+        gs.mesh.get(&topic_hashes[0]).unwrap().len(),
+        4,
+        "topic1 should have mesh_n 4 peers after re-subscribe"
+    );
+}
+
+/// Test fanout behavior with topic-specific configuration
+#[test]
+fn test_fanout_with_topic_config() {
+    let topic = String::from("topic1");
+    let topic_hash = TopicHash::from_raw(topic.clone());
+
+    let topic_config = TopicMeshConfig {
+        mesh_n: 4,
+        mesh_n_low: 2,
+        mesh_n_high: 7,
+        mesh_outbound_min: 1,
+    };
+
+    // turn off flood publish to test fanout behaviour
+    let config = ConfigBuilder::default()
+        .flood_publish(false)
+        .set_topic_config(topic_hash.clone(), topic_config)
+        .build()
+        .unwrap();
+
+    let (mut gs, _, receivers, topic_hashes) = inject_nodes1()
+        .peer_no(10) // More than mesh_n
+        .topics(vec![topic.clone()])
+        .to_subscribe(true)
+        .gs_config(config)
+        .create_network();
+
+    assert!(
+        gs.unsubscribe(&Topic::new(topic.clone())),
+        "Should unsubscribe successfully"
+    );
+
+    let publish_data = vec![0; 42];
+    gs.publish(Topic::new(topic), publish_data).unwrap();
+
+    // Check that fanout size matches the topic-specific mesh_n
+    assert_eq!(
+        gs.fanout.get(&topic_hashes[0]).unwrap().len(),
+        4,
+        "Fanout should contain topic-specific mesh_n 4 peers for this topic"
+    );
+
+    // Collect publish messages
+    let publishes = receivers
+        .into_values()
+        .fold(vec![], |mut collected_publish, c| {
+            let priority = c.priority.get_ref();
+            while !priority.is_empty() {
+                if let Ok(RpcOut::Publish { message, .. }) = priority.try_recv() {
+                    collected_publish.push(message);
+                }
+            }
+            collected_publish
+        });
+
+    // Verify sent to topic-specific mesh_n number of peers
+    assert_eq!(
+        publishes.len(),
+        4,
+        "Should send a publish message to topic-specific mesh_n 4 fanout peers"
+    );
+}
```

Comment:
thanks for adding these tests!
Can we also add a test for the happy and unhappy paths of codec decoding, so that we assert `MessageSizeTooLargeForTopic` works? this relates with my comment on the `protocol.rs`